### PR TITLE
Disabled validators runtime API

### PR DESCRIPTION
**Pull request by tdimitrov**
Exposes disabled validators list via a runtime API.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3550779

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3553002

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3638774

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3716470

File: polkadot/primitives/src/runtime_api.rs [hunk 1]
```diff
@@ -257,5 +258,11 @@ sp_api::decl_runtime_apis! {
 		/// Returns candidate's acceptance limitations for asynchronous backing for a relay parent.
 		#[api_version(7)]
 		fn async_backing_params() -> AsyncBackingParams;
+
+		/***** Added in v8 *****/
+
+		/// Returns a sorted Vec with the `ValidatorIndex` of all disabled validators.
```
Comment (ordian) [inline polkadot/primitives/src/runtime_api.rs:264]:
```suggestion
		/// Returns a list of all disabled validators at the given block.
```

File: polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs [hunk 2]
```diff
@@ -15,3 +15,30 @@
...
+	// mapping from raw validator index to `ValidatorIndex`
+	// this computation is the same within a session, but should be cheap
+	let reverse_index = shuffled_indices
+		.iter()
+		.enumerate()
+		.map(|(i, v)| (v.0, ValidatorIndex(i as u32)))
+		.collect::<BTreeMap<u32, ValidatorIndex>>();
+
+	// we might have disabled validators who are not parachain validators
+	<pallet_session::Pallet<T>>::disabled_validators()
```
Comment (BradleyOlson64) [inline polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs:40]: I think I get what's going on here, but if you could check my thinking that would be awesome.

With inputs:
`shuffled_indices = [6, 9, 1]`
`disabled_validators = [9, 6]`

We'd get the steps

reverse_index = BTreeMap with entries (6, 1), (9, 2), (1, 3)

returned value = vec![2, 1]

**Review by BradleyOlson64 — APPROVED** [review]
Looks good to me 👍

File: polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs (hunk 2 above)
Comment (ordian) [reply polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs:40]: Almost, the only correction is that indices are 0-based, not 1-based.

File: polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs [hunk 3]
```diff
@@ -16,18 +16,28 @@
...
-	// 	.map(|v| ValidatorIndex(v))
-	// 	.collect()
-	// TODO
-	Vec::new()
+pub fn disabled_validators<T>() -> Vec<ValidatorIndex>
+where
+	T: pallet_session::Config + shared::Config,
+{
+	let shuffled_indices = <shared::Pallet<T>>::active_validator_indices();
+	// mapping from raw validator index to `ValidatorIndex`
```
Comment (tdimitrov) [inline polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs:31]: `ValidatorIndex` in polkadot is an index in active validators array while [`ValidatorIndex` in substrate](https://github.com/paritytech/polkadot-sdk/blob/5174b9d2d7ae67ec05a481508e7ec275e4d4f1d9/polkadot/runtime/parachains/src/shared.rs#L143) is completely different index in `all_validators` ?

Or put in other way:
1. [0....N] - all validator indecies (in substrate)
2. The array above gets shuffled and truncated to become `active_validator_indices` which is still 'substrate indexes' [here](https://github.com/paritytech/polkadot-sdk/blob/5174b9d2d7ae67ec05a481508e7ec275e4d4f1d9/polkadot/runtime/parachains/src/shared.rs#L194))
3. [`ValidatorIndex` in polkadot](https://github.com/paritytech/polkadot-sdk/blob/7cbe0c76ef8fd2aabf9f07de0156941ce3ed44b0/polkadot/primitives/src/v6/mod.rs#L115) is an index within `active_validator_indecies`. It is one abstraction above the substrate indecies.

Is this correct?

File: polkadot/node/core/runtime-api/src/cache.rs [hunk 4]
```diff
@@ -444,6 +446,21 @@ impl RequestResultCache {
 		self.minimum_backing_votes.insert(session_index, minimum_backing_votes);
 	}
 
+	pub(crate) fn disabled_validators(
+		&mut self,
+		relay_parent: &Hash,
+	) -> Option<&Vec<ValidatorIndex>> {
+		self.disabled_validators.get(relay_parent).map(|v| &*v)
```
Comment (AndreiEres) [inline polkadot/node/core/runtime-api/src/cache.rs:453]: Out of curiosity, what that `.map(|v| &*v)` does?

**Review by AndreiEres — APPROVED** [review]
lgtm

File: polkadot/node/core/runtime-api/src/cache.rs (hunk 4 above)
Comment (ordian) [reply polkadot/node/core/runtime-api/src/cache.rs:453]: It converts &mut to &

File: polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs (hunk 3 above)
Comment (ordian) [reply polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs:31]: Yes. What is confusing is that `active_validator_indices` holds a vec of `ValidatorIndex`, whereas it should be some other type representing raw substrate index. We use `ValidatorIndex` to represent an index in that vec, so it's in range of 0..max_validators.

Comment (tdimitrov) [reply polkadot/runtime/parachains/src/runtime_api_impl/vstaging.rs:31]: Oh I see... Thanks for the explanation.

File: cumulus/client/relay-chain-rpc-interface/src/rpc_client.rs [hunk 5]
```diff
@@ -597,6 +597,14 @@ impl RelayChainRpcClient {
 			.await
 	}
 
+	pub async fn parachain_host_disabled_validators(
```
[the remark this answers is not in the document]
Comment (ordian) [reply cumulus/client/relay-chain-rpc-interface/src/rpc_client.rs:600]: https://github.com/paritytech/cumulus/pull/2160#issuecomment-1562490741

Well, I agree we should limit the scope for collators. But this is out of scope (hehe) of this PR.

**Comment by tdimitrov** [comment]
@eskimor @ordian should we merge this?

**Comment by ordian** [comment]
no objection from me

**Pull request merged**